// ngx-quicklink/esm2022/lib/quicklink.module.mjs
import { NgModule } from '@angular/core';
import { ObservableLinkHandler, PreloadLinkHandler, LinkHandler, } from './link-handler.service';
import { LinkDirective } from './link.directive';
import { PrefetchRegistry } from './prefetch-registry.service';
import { QuicklinkStrategy } from './quicklink-strategy.service';
import * as i0 from "@angular/core";
export const quicklinkProviders = [
    {
        provide: LinkHandler,
        useClass: ObservableLinkHandler,
        multi: true,
    },
    {
        provide: LinkHandler,
        useClass: PreloadLinkHandler,
        multi: true,
    },
    PrefetchRegistry,
    QuicklinkStrategy,
];
export class QuicklinkModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: QuicklinkModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: QuicklinkModule, imports: [LinkDirective], exports: [LinkDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: QuicklinkModule, providers: quicklinkProviders }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: QuicklinkModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [LinkDirective],
                    exports: [LinkDirective],
                    providers: quicklinkProviders,
                }]
        }] });
